tests/test_core.py

```diff
@@ -4044,6 +4044,111 @@ def test_dylink_hyper_dupe(self):
       full = run_js('src.cpp.o.js', engine=JS_ENGINES[0], full_output=True, stderr=STDOUT)
       self.assertContained("warning: trying to dynamically load symbol '_sideg' (from 'third%s') that already exists" % dylib_suffix, full)
 
+  @needs_dlfcn
+  def test_dylink_dso_needed(self):
+    # test that linking to shared library B, which is linked to A, loads A as well.
+    # main is also linked to C, which is also linked to A. A is loaded/initialized only once.
+    #
+    #          B
+    #   main <   > A
+    #          C
+    open('liba.cpp', 'w').write(r'''
```

Comment:
Yes, sure. There is even a potential problem with how it is presently written - as it potentially leaks an fd and implicitly relies on GC to finalize/close the opened file. I kept it that way because all the code in `test_core.py` does it that way.

If we are to change something here, then what you suggest would work

```py
    with open('liba.cpp', 'w') as f:
        f.write(r'''
            ...
```

but it is 2 lines and more noise than it should be. I thus suggest to add the following small helpers:

```py
# readfile reads data from file @path.
def readfile(path):
    with open(path, "r") as f:
        return f.read()

# writefile writes data to file @path.
def writefile(path, data):
    with open(path, "w") as f:
        f.write(data)
```

and switch to using them universally, e.g.:

```py
    writefile('liba.cpp', r'''
        ...
```

The switch should be then better done globally in one go.

--------

The above is of course minor and for now I change the patch to use `with open...`.
(I keep the style to be old and consistent with the around in browser.test_dynamic_link)

Reply:
I'm been transitioning the test code over to "with open().." for the last for months.  I agree we should probably do a wholesale switch at some point.    In the mean time I'd rather see "with open.." used in new code.

In terms on the long term approach, AFAICT using "with open.." only as a single line extra (not two as you claim), but when we do that large refactor I'd be OK with a helper function I think.

Reply:
ok. (the patch now uses "with open()..." for new test)